### ISSUE TITLE
Nested input types generate an error at introspection time

### ISSUE THREAD

**Issue by djc** [async-graphql/async-graphql#880]
I have some types like this:

```rust
#[derive(InputObject)]
pub(super) struct EditDomainInput {
    domain: String,
    title: Option<String>,
    name: Option<String>,
    description: Option<String>,
    logo: Option<String>,
    contact_info: Option<ContactInfo>,
}

#[derive(InputObject)]
pub struct ContactInfo {
    address: Option<Address>,
    phone_number: Option<String>,
    payment_url: Option<String>,
    contact_email: Option<String>,
}

#[derive(InputObject)]
pub struct Address {
    country: Option<String>,
    region: Option<String>,
    city: Option<String>,
    postal_code: Option<String>,
    street_address: Option<String>,
}
```

This compiles fine and I can generate SDL  from it.

## Expected Behavior

Introspection queries against this schema should work.

## Actual Behavior

Introspection query fails; error from GraphiQL:

> `Introspection must provide input type for arguments, but received: ContactInfo.`

## Specifications

- Version: 3.0.36
- Platform: macOS
- Subsystem: input  types

**Comment by sunli829** [comment]
I can't reproduce it, can you provide a repository that reproduces this issue? 😁

**Comment by djc** [comment]
Here's a full reproducer: https://github.com/djc/nested-input.

What I find really surprising about this is that introspection works okay as long as I comment out the `_input` argument to the `_edit_foo()` mutation.

**Comment by sunli829** [comment]
`Info` is both an input type and an output type causing this error.

```rust
#[derive(InputObject)]
#[graphql(name = "InfoInput")] // Add this attribute to rename the input type name
pub struct Info {
    email: Option<String>,
}
```

**Comment by sunli829** [comment]
This error cannot currently be reported at compile time, I will add a runtime check. 🙂

**Comment by djc** [comment]
Why can't Rust types implement both the input and the output traits?

**Comment by sunli829** [comment]
This is not allowed in the GraphQL spec, they need to have different names.

**Comment by sunli829** [comment]
You can implement as input object and output object for the same Rust type, but for GraphQL they are actually different objects, so must have different names as well.

**Comment by djc** [comment]
Would be nice if there was a way to do this similar to the way currently multiple names can be declared depending on any generic type parameters.

**Comment by sunli829** [comment]
This should not be possible. 😂

**Comment by sunli829** [comment]
Or I don't know how to do it. 😄

**Comment by sunli829** [comment]
Yes, procedural macros can do it.

**Comment by sunli829** [comment]
You mean, always automatically add the `Input` suffix?

**Comment by djc** [comment]
Ah, sorry -- I didn't actually mean that the macros should implicitly add a suffix to input type names. I meant that there should be a way for me to reuse the same type under a different name, but it looks like that already exists.

I think it might actually be nice if there was a way I could opt in to add `Input` suffixes to all input type names on a per-crate basis to prevent this issue from happening, but I could understand that you might not want to enable this by default.

**Comment by sunli829** [comment]
Well, I'm now going to add a runtime check that panics when the same Rust type has the same name.

**Comment by sunli829** [comment]
Released in `v3.0.37`
